cedar-policy-validator/src/custom_schema/ast.rs

```diff
@@ -0,0 +1,299 @@
+use cedar_policy_core::{
+    ast::Id,
+    parser::{
+        cst::{Name, Ref},
+        Loc, Node,
+    },
+};
+use itertools::Either;
+use itertools::Itertools;
+use nonempty::NonEmpty;
+use smol_str::SmolStr;
+
+use crate::SchemaTypeVariant;
+
+const IPADDR_EXTENSION: &str = "ipaddr";
+const DECIMAL_EXTENSION: &str = "decimal";
+pub const EXTENSIONS: [&str; 2] = [IPADDR_EXTENSION, DECIMAL_EXTENSION];
+pub const BUILTIN_TYPES: [&str; 3] = ["Long", "String", "Bool"];
+
+pub(super) const CEDAR_NAMESPACE: &str = "__cedar";
+
+pub type IdentOrString = Either<Node<Id>, Node<SmolStr>>;
+pub type Schema = Vec<Node<Namespace>>;
+
+/// A path is a non empty list of identifiers that forms a namespace + type
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct Path(pub Node<NonEmpty<Id>>);
+
+impl Path {
+    /// Create a [`Path`] with a single entry
+    pub fn single(node: Id, loc: Loc) -> Self {
+        Self(Node::with_source_loc(NonEmpty::new(node), loc))
+    }
+
+    /// Create [`Path`] with a head and an iterator
+    pub fn new(head: Id, rest: impl IntoIterator<Item = Id>, loc: Loc) -> Self {
+        let tail = rest.into_iter().collect();
+        Self(Node::with_source_loc(NonEmpty { head, tail }, loc))
+    }
+
+    /// Borrowed iteration of the [`Path`]'s elements
+    pub fn iter(&self) -> impl Iterator<Item = &Id> {
+        self.0.node.iter()
+    }
+
+    /// Source [`Loc`] of this [`Path`]
+    pub fn loc(&self) -> &Loc {
+        &self.0.loc
+    }
+
+    /// Consume the [`Path`] and get an owned iterator over the elements
+    pub fn into_iter(self) -> impl Iterator<Item = Node<Id>> {
+        let loc = self.0.loc;
+        self.0
+            .node
+            .into_iter()
+            .map(move |x| Node::with_source_loc(x, loc.clone()))
+    }
+
+    /// Get the base type name as well as the (potentially empty) prefix
+    pub fn split_last(self) -> (Vec<Id>, Id) {
+        let ne = self.0.node;
+        let first = ne.head;
+        let mut rest = ne.tail;
+        match rest.pop() {
+            Some(last) => {
+                rest.insert(0, first);
+                (rest, last)
+            }
+            None => (rest, first),
+        }
+    }
+}
+
+impl std::fmt::Display for Path {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let s = self.0.node.iter().map(|id| id.as_ref()).join("::");
+        write!(f, "{s}")
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct QualName {
+    pub path: Option<Path>,
+    pub id: SmolStr,
+}
```

Comment:
two comments

1) I thought `Path` already was an optional number of namespaces, plus a final `id`
2) `id` is a confusing name for a field of type `SmolStr`.  is it an identifier or an arbitrary string?

basically I'm unsure at a glance whether this represents a (potentially qualified) typename, or an EntityUid

Reply:
EntityUId. A path is a nonempty set of namespaces, but they all have to be `id`s. An actions entity uid can be any string